pkg/scheduler/scheduler.go

```diff
@@ -32,6 +32,7 @@ import (
 	coreinformers "k8s.io/client-go/informers/core/v1"
 	clientset "k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/cache"
+	clientretry "k8s.io/client-go/util/retry"
```

Comment:
> I'm quite confused here.
> 
> > A race condition occurs because the medium priority pod is still attempting to schedule at the same time when removeNominatedNodeName() is being called to clear it's nominated node value.
> 
> Is "still attempting to schedule" referring to binding? Otherwise there shouldn't be 2 pods being scheduled at the same time.

Yes, sorry.  I think the go-routine from `scheduleOne` that binds the pod is still running asynchronously when the high-priority pod preempts medium-priority pod.

In `scheduler.preempt()`, this line gets the nominated pods to clear:
```
node, victims, nominatedPodsToClear, err := sched.Algorithm.Preempt(ctx, prof, state, preemptor, scheduleErr)
```
but it is *possible* (although unlikely, which is why it shows up as a test flake) that nominatedPodsToClear gets out of date before it is updated later by the call to `sched.podPreemptor.removeNominatedNodeName(p)`.

The integration testing logs I looked at show response of 409 coming back when it tries to update the pod status, so there is definitely something causing a conflict.  Retrieving the latest and retrying the update succeeds.

Whether there is another way to solve this, I'm not sure. I was thinking at the very least, if a 409 comes back when the scheduler is trying to update a pod, it should retry.  Currently, if any failure occurs during removeNominatedNodeName, it only logs the error and continues to return successfully.  That is actually what causes the timeout to occur in the integration test... it doesn't know that it failed and is waiting for the nominated node name to be cleared from the medium-priority pod, but it will never occur at that point.

> > Often times this test succeeds, but if the timing is unlucky, then the medium-priority pod will never have its nominated node name removed because it didn't retry on conflict,
> 
> What's the conflict exactly? I would say that, if there is a conflict, the whole scheduling attempt should be retried. Perhaps this is not a non-critical error:
> 
> https://github.com/kubernetes/kubernetes/blob/a15168288792daa23c993e551894229161dabdd0/pkg/scheduler/scheduler.go#L452-L456
> 
> And if you disagree with my statement above, let's use informer's cache as #90660

I don't know if this is a critical error or not. I think it really only happens when you have two successive preemptions on the same node within a very short period of time. For practical purposes, this is probably an edge case, but I guess a system under heavy load could experience this. 

Regarding retrying to whole scheduling attempt. The problem I see with that is the conflict occurs when it is trying to remove the nominated node name from another pod and it seems like the preemption has already occurred, so it might be hard to retry the whole thing.

I'm definitely new to the scheduler, so please take what I'm saying with a grain of salt and I will defer to those more familiar with it to let me know if I am wrong about something.

Thanks, I really do appreciate the comments and questions.

Reply:
One more thing, regarding the `// We do not return as this error is not critical. ` comment. Read the comment above that for the assumption being made:
```
        // Clearing nominated pods should happen outside of "if node != nil". Node could
	// be nil when a pod with nominated node name is eligible to preempt again,
	// but preemption logic does not find any node for it. In that case Preempt()
	// function of generic_scheduler.go returns the pod itself for removal of
	// the 'NominatedNodeName' field.
	for _, p := range nominatedPodsToClear {
		rErr := sched.podPreemptor.removeNominatedNodeName(p)
		if rErr != nil {
			klog.Errorf("Cannot remove 'NominatedNodeName' field of pod: %v", rErr)
			// We do not return as this error is not critical.
		}
	}
```

It is assuming this happens because the node is nil, but `rErr` is not checked, and it is not because the node is nil, but a conflict occurs, which I believe to be a different problem that should not be ignored.

Reply:
> nit: no need for the alias

Fixed, thanks